### PR TITLE
High Scalability of metastore

### PR DESCRIPTION
**Pull request by NishitAtlan**
## Change description

> Description here

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#META-3009]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: common/src/main/java/org/apache/atlas/ha/HAConfiguration.java [hunk 1]
```diff
@@ -74,6 +74,15 @@ public static boolean isHAEnabled(Configuration configuration) {
         return ret;
     }
 
+    /**
+     * Return whether HS (Active-Active HA) is enabled or not.
+     * @param configuration underlying configuration instance
+     * @return
+     */
+    public static boolean isActiveActiveHAEnabled(Configuration configuration) {
```
Comment (mehtaanshul) [inline common/src/main/java/org/apache/atlas/ha/HAConfiguration.java:82]: Change to isHSEnabled

File: repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java [hunk 2]
```diff
@@ -63,14 +65,16 @@ public class IndexRecoveryService implements Service, ActiveStateChangeHandler {
...
+    public IndexRecoveryService(Configuration config, AtlasGraph graph, ICuratorFactory curatorFactory) {
         this.configuration               = config;
         this.isIndexRecoveryEnabled      = config.getBoolean(ApplicationProperties.INDEX_RECOVERY_CONF, DEFAULT_INDEX_RECOVERY);
         long recoveryStartTimeFromConfig = getRecoveryStartTimeFromConfig(config);
         long healthCheckFrequencyMillis  = config.getLong(SOLR_STATUS_CHECK_RETRY_INTERVAL, SOLR_STATUS_RETRY_DEFAULT_MS);
         this.recoveryInfoManagement      = new RecoveryInfoManagement(graph);
 
-        this.recoveryThread = new RecoveryThread(recoveryInfoManagement, graph, recoveryStartTimeFromConfig, healthCheckFrequencyMillis);
+        final InterProcessMutex indexRecoveryLock = curatorFactory.lockInstanceWithDefaultZkRoot(INDEX_RECOVERY_LOCK);
+        final boolean isActiveActiveHAEnabled = HAConfiguration.isActiveActiveHAEnabled(configuration);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:76]: isHSEnabled

File: repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java [hunk 3]
```diff
@@ -63,14 +65,16 @@ public class IndexRecoveryService implements Service, ActiveStateChangeHandler {
...
-    public IndexRecoveryService(Configuration config, AtlasGraph graph) {
+    public IndexRecoveryService(Configuration config, AtlasGraph graph, ICuratorFactory curatorFactory) {
         this.configuration               = config;
         this.isIndexRecoveryEnabled      = config.getBoolean(ApplicationProperties.INDEX_RECOVERY_CONF, DEFAULT_INDEX_RECOVERY);
         long recoveryStartTimeFromConfig = getRecoveryStartTimeFromConfig(config);
         long healthCheckFrequencyMillis  = config.getLong(SOLR_STATUS_CHECK_RETRY_INTERVAL, SOLR_STATUS_RETRY_DEFAULT_MS);
         this.recoveryInfoManagement      = new RecoveryInfoManagement(graph);
 
-        this.recoveryThread = new RecoveryThread(recoveryInfoManagement, graph, recoveryStartTimeFromConfig, healthCheckFrequencyMillis);
+        final InterProcessMutex indexRecoveryLock = curatorFactory.lockInstanceWithDefaultZkRoot(INDEX_RECOVERY_LOCK);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:75]: Here we are calling this method without check activeactiveenable
In tasks, we are calling this just before acquiring lock and only if activeactiveenabled
and in typedef we are using a slightly diff method.

File: webapp/src/main/java/org/apache/atlas/web/rest/TypeCacheRefreshREST.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     * @HTTP 500 if there is an error refreshing type def cache
+     */
+    @POST
+    @Path("/refresh")
+    @Timed
+    public void refreshCache(@QueryParam("expectedFieldKeys") int expectedFieldKeys) throws AtlasBaseException, RepositoryException, InterruptedException {
+        try {
+            refreshTypeDef(expectedFieldKeys);
+        } catch (Exception e) {
+            LOG.error(e.getMessage(),e);
```
Comment (mehtaanshul) [inline webapp/src/main/java/org/apache/atlas/web/rest/TypeCacheRefreshREST.java:59]: Let's improve all the error logs by adding a custom message, so we can search on sentry.
Also our Rest APIsonly return AtlasBaseException -> convert other exceptions to this exception for consistency with a proper exception message

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java [hunk 5]
```diff
@@ -88,41 +87,66 @@ public void run() {
...
+
+    private void releaseLock(InterProcessMutex lock) {
+        try {
+            if (isActiveActiveHAEnabled && lock != null && lock.isOwnedByCurrentThread()) {
+                LOG.info("About to release task lock");
+                lock.release();
+                LOG.info("successfully released task lock");
+            }
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:134]: Let's improve all the error logs by adding a custom message, so we can search on sentry.

File: repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java [hunk 6]
```diff
@@ -248,6 +265,18 @@ private void stopMonitoring() {
...
+
+        private void releaseLock(InterProcessMutex lock) {
+            try {
+                if (isActiveActiveHAEnabled && lock != null && lock.isOwnedByCurrentThread()) {
+                    LOG.info("About to release index recovery lock");
+                    lock.release();
+                    LOG.info("successfully released index recovery lock");
+                }
+            } catch (Exception e) {
+                LOG.error(e.getMessage(), e);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:277]: Let's improve all the error logs by adding a custom message, so we can search on sentry.

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java [hunk 7]
```diff
@@ -68,18 +72,13 @@ public void run() {
...
             try {
-                if (latch != null && latch.getCount() != 0) {
-                    LOG.info("TaskQueueWatcher: Waiting on Latch, current count: {}", latch.getCount());
-                    latch.await();
-                }
-
-                if (latch != null) {
-                    LOG.info("TaskQueueWatcher: Latch wait complete!!");
+                if (isActiveActiveHAEnabled) {
+                    lock = acquireDistributedLock();
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:80]: Asking out of curiosity what happens if the lock is acquired by another node, does code remain stuck here until other pod releases lock?

File: repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java (hunk 3 above)
Comment (mehtaanshul) [reply repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:75]: Please review this and let's discuss if possible

Comment (NishitAtlan) [reply repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:75]: Addressed this by making all the places consistent

File: webapp/src/main/java/org/apache/atlas/web/rest/TypeCacheRefreshREST.java (hunk 4 above)
Comment (NishitAtlan) [reply webapp/src/main/java/org/apache/atlas/web/rest/TypeCacheRefreshREST.java:59]: This is addressed

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java (hunk 5 above)
Comment (NishitAtlan) [reply repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:134]: This is addressed

File: repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java (hunk 6 above)
Comment (NishitAtlan) [reply repository/src/main/java/org/apache/atlas/repository/graph/IndexRecoveryService.java:277]: This is addressed

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java (hunk 7 above)
Comment (NishitAtlan) [reply repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:80]: Yes

**Pull request merged**